src/refcount.rs

```diff
@@ -65,39 +80,55 @@ pub trait RefCounter: Sealed + Clone + Unpin + Send + Sync + 'static {
     fn into_either(self) -> Either;
 
     #[doc(hidden)]
-    fn as_ptr(&self) -> *const AtomicBool;
+    fn as_ptr(&self) -> *const (AtomicBool, DropNotice);
+
+    /// Returns a `DropNotice` that resolves once this address becomes disconnected.
+    fn disconnect_notice(&self) -> DropNotice;
 }
 
 impl RefCounter for Strong {
     fn is_connected(&self) -> bool {
-        self.strong_count() > 0 && self.0.load(Ordering::Acquire)
+        self.strong_count() > 0 && self.0 .0.load(Ordering::Acquire)
     }
 
     fn is_last_strong(&self) -> bool {
+        let _lock = self.1.read().unwrap();
         Arc::strong_count(&self.0) == 1
     }
 
     fn strong_count(&self) -> usize {
+        let _lock = self.1.read().unwrap();
         Arc::strong_count(&self.0)
     }
 
     fn into_either(self) -> Either {
         Either::Strong(self)
     }
 
-    fn as_ptr(&self) -> *const AtomicBool {
+    fn as_ptr(&self) -> *const (AtomicBool, DropNotice) {
         Arc::as_ptr(&self.0)
     }
+
+    fn disconnect_notice(&self) -> DropNotice {
+        self.0 .1.clone()
+    }
 }
 
 impl RefCounter for Weak {
     fn is_connected(&self) -> bool {
+        let lock = self.1.write().unwrap();
```

Comment:
I wonder if it'd be possible to just use a normal bool that would be inside of the RwLock? Or would that also be susceptible to the same race condition?

Reply:
Not sure I follow. The `RwLock` actually only contains a `()`, not the main counting `Arc`.

Reply:
It's probably not important - I don't really remember what I meant with this.